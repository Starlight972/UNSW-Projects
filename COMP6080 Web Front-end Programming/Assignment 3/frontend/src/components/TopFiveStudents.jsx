import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

// Own components
import StudentPoints from './StudentPoints'

const TopFiveStudents = ({ questions, results }) => {
  const [points, setPoints] = React.useState([]);
  const [playerPoints, setPlayerPoints] = React.useState([]);
  const [topFive, setTopFive] = React.useState([]);

  // get each question's point
  React.useEffect(() => {
    const tempPoints = [];
    for (const question of questions) {
      tempPoints.push(question.points)
    }
    setPoints(tempPoints);
  }, [questions])

  // calculate each player's total points
  React.useEffect(() => {
    const tempPlayerPoints = [];
    let player = {
      name: '',
      points: 0
    }
    for (const result of results) {
      player = { ...player, name: result.name }
      const answers = result.answers;
      let totalPoints = 0;
      for (let i = 0; i < answers.length; i++) {
        answers[i].correct && (totalPoints += points[i])
      }
      player = { ...player, points: totalPoints }
      tempPlayerPoints.push(player);
    }
    setPlayerPoints(tempPlayerPoints);
  }, [points])
  // console.log(playerPoints);

  // Find Top 5 students
  React.useEffect(() => {
    let totalPoints = [];
    for (const player of playerPoints) {
      totalPoints.push(player.points);
    }
    const tempTopFive = [];
    let student = {
      name: '',
      points: 0
    }

    for (let i = 0; i < 5; i++) {
      console.log(totalPoints);
      let max = -1;
      totalPoints.forEach((element) => {
        if (max < element) {
          max = element
        }
      })
      totalPoints = totalPoints.filter(item => item !== parseInt(max));
      student = { ...student, points: max }
      const tempName = playerPoints.filter(player => player.points === max);
      if (tempName.length > 0) {
        student = { ...student, name: tempName[0].name, points: max }
        // console.log(student);
        tempTopFive.push(student);
        // console.log(tempTopFive);
      }
    }
    setTopFive(tempTopFive);
  }, [playerPoints])

  // console.log(topFive);
  return (
    <Wrapper>
      <Title>Top 5</Title>
      {topFive.length > 0 && topFive.map((item, index) => (
        <StudentPoints key={index} name={item.name} points={item.points}/>
      ))}
    </Wrapper>
  )
}

export default TopFiveStudents;

TopFiveStudents.propTypes = {
  questions: PropTypes.any,
  results: PropTypes.any
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
  margin-left: 4rem;
`;

const Title = styled.h1`
  font-size: 3rem;
  color: #3AA3A0;
  margin: auto;
  margin-top: 3rem;
`;
