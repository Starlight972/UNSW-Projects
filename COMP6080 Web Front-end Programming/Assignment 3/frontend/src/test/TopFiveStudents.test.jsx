import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';

// own component
import TopFiveStudents from '../components/TopFiveStudents';

describe('<TopFiveStudents />', () => {
  let wrapper;

  it('renders without crashing', () => {
    wrapper = shallow(<TopFiveStudents />);
  })

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

test('renders text elements', () => {
  render(<TopFiveStudents questions={[]} results={[]} />);
  const text = screen.getByText('Top 5');
  expect(text).toBeInTheDocument();
});
