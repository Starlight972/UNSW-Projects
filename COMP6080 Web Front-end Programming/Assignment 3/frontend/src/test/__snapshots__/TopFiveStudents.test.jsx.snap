// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopFiveStudents /> matches snapshot 1`] = `ShallowWrapper {}`;
